### ISSUE TITLE
Any chance that this will ever connect to my Visual Studio Github account?

### ISSUE THREAD

**Issue by cyberangel67** [github-for-unity/Unity#909]
I just came across this and wanted to know if there is any chance that at some point in the future, this may include connecting to my Visual Studio Github repository.

**Comment by StanleyGoldman** [comment]
Hey @cyberangel67. Thanks for checking our stuff out.

What do you mean by "Visual Studio GitHub repository"?

**Comment by cyberangel67** [comment]
By being able to connect to Github on my VisualStudio.com account.

**Comment by StanleyGoldman** [comment]
Can you give me a link to describe what you are referring to?

😅💦  I'm struggling over here..

**Comment by cyberangel67** [comment]
All I know is that my Microsoft Account when connect in Visual Studio I can connect to

username.visualstudio.com

And have repositories for all my projects. This is I think under their Team Services umbrella. Where I can sync my code to a github repository there.

**Comment by cyberangel67** [comment]
Maybe this will help

https://azure.microsoft.com/en-gb/services/devops/repos/

**Comment by StanleyGoldman** [comment]
OH!... I understand what you mean now.
**TLDR**: Yes

### Long version
You threw me off the scent with...
> connecting to my Visual Studio Github repository.

**Git** is a revision control system, a tool to manage your source code history.
**GitHub** is a hosting service for **Git** repositories.

So what you should have said is...
> connecting to my Visual Studio Git repository.

And I would've understood you immediately...

This is not entirely your fault...

**GitHub for Unity** is a **Git** client for Unity with a few special features for **GitHub**. 😭 
... But it's called **GitHub for Unity** cause it's being made by us at **GitHub** 😭 😭

So yes, it is just a **Git** client and you can use it with Visual Studio's git services...

Sheesh...

**Comment by cyberangel67** [comment]
But it doesn't connect, so if it works with the Visual Studio Git repository, how does is set up because very time I try it won't authenticate.

**Comment by cyberangel67** [comment]
wow, so it doesn't actually work and you close it.

**Comment by StanleyGoldman** [comment]
It's not a problem to reopen the issue.
I closed the issue because I thought I answered the question.

We are happy to help if you have more questions...

The "Login" and "Publish"  functions are specific to GitHub users only.

If you already have Git in your project, all the other functions (commit, fetch, push, pull, etc) should just work for you.

**Comment by StanleyGoldman** [comment]
Here are some instructions I've given to a user previously to help him out...

> Existing Project
> 1. Make sure you can perform git operations on the command line to your chosen provider.
> 2. Clone it on the command l ine
> 3. Open it Unity & install GitHub for Unity
> 4. Fetch/Push/Pull should just work
> 
> New Project
> 1. Make sure you can perform git operations on the command line to your chosen provider.
> 2. Open a new Unity Project and install GitHub for Unity.
> 3. Create your repo with your provider, Save the git url in Settings in GitHub for Unity
> 4. Fetch/Push/Pull should just work

**Comment by cyberangel67** [comment]
Ok, now we are right back to my initial question.

Can we add Github for Unity to actually connect to Visual Studio Github repository

I am not comfortable with command line tools, in fact I detest them.

So is there a reason why, this won't be considered?

**Comment by StanleyGoldman** [comment]
Unfortunately we do not have plans to support any other providers APIs.

You don't have to use the command line.
It sounds like you are already using **_something_** to work your Git repositories locally.
Are you able to create, clone and commit to a project with those tools?

If you using GitHub for Unity, it would just complement those tools that you are already using.

**Comment by cyberangel67** [comment]
I use Visual Studio normally for projects, but as Unity hides most of the files from VS I can't commit changes to files that aren't shown. Nor can I revert files not shown.

Well at least I don't think I can, cause I can't connect a Unity project to my remote repo.

**Comment by StanleyGoldman** [comment]
Okay, How about we start from the beginning...

- Where is the solution stored on your computer?
- Where is the git database stored on your computer?
- What IDE do you start with to look at your work?

> I use Visual Studio normally for projects

Do you mean you use Visual Studio and Visual Studio's git tools normally on Visual Studio solutions? As opposed to Unity solutions?

> Unity hides most of the files from VS I can't commit changes to files that aren't shown. Nor can I revert files not shown.

What files are you talking about?

Also.. if we are going to make this work, we have to agree on some things.
You have said "Visual Studio **Github** repository" a few times now, but there is no such thing. It is a Visual Studio **Git** repository.

Something to consider is, until you learn how to use Git better, it may be easier for you to use a Git repository on GitHub instead of Visual Studio.

**Comment by cyberangel67** [comment]
> Okay, How about we start from the beginning...
> 
> * Where is the solution stored on your computer?

The solution is just a normal Unity project, saved one of my drives.

> * Where is the `.git` folder (the git database) stored on your computer?

There is no git database for my Unity projects.

> * What IDE do you start with to look at your work?

I use Unity Editor for Unity, Visual Studio to edit the source files.

> > I use Visual Studio normally for projects
> 
> Do you mean you use Visual Studio and Visual Studio's git tools normally on Visual Studio solutions? As opposed to Unity solutions?

That is correct.

> > Unity hides most of the files from VS I can't commit changes to files that aren't shown. Nor can I revert files not shown.
> 
> What files are you talking about?

The entire Unity project, when you go into Visual Studio the only source files exposed are the C# files.

> Also.. if we are going to make this work, we have to agree on some things.
> You have said "Visual Studio **Github** repository" a few times now, but there is no such thing. It is a Visual Studio **Git** repository.

Isn't

"Visual Studio **Github** repository"

the same as

Visual Studio **Git** repository

> Something to consider is, until you learn how to use Git better, it may be easier for you to use a Git repository on GitHub instead of Visual Studio.

I am not using a public repository, I already have a private repository with VS Team Services. Where I would like to connect to from my Unity Editor. I am very comfortable using VS to connect my VS Services to do source control. Just Unity projects can't be done in this manner.

Now as this is the first time I am looking for a decent solution for Unity, I came across this a few days ago. And was extremely disappointed that I can only connect to a public github repository. As these projects are closed source, I am not going to stick this code in something that is public.

**Comment by StanleyGoldman** [comment]
This tool **does** work with other **Git** providers than **GitHub**. _(Also GitHub has private repositories as well)_. Also, I was not suggesting you put your private project somewhere public, I was suggesting you play with the tools in a sample project, outside of your project, to get a better idea of how they work.

I was able to create a repository in Unity, create a git repo with GitHub for Unity, open the solution in Visual Studio, make a commit from Visual Studio, and publish it to visualstudio.com. As you can see here...

![image](https://user-images.githubusercontent.com/417571/45691043-bb93a680-bb25-11e8-8d7d-e21e677a41d7.png)

![image](https://user-images.githubusercontent.com/417571/45691266-42e11a00-bb26-11e8-8400-6fa47554a6b5.png)

Here is what I did.
1. Created a new project in Unity.
1. Installed GitHub for Unity.
1. Initialized a repository from GitHub for Unity.
1. Right click the Asset Explorer -> Open C# Project
1. Ran Sync in Visual Studio's Team Explorer, publishing the project to visualstudio.com
1. Created a new Behavior script in Unity
1. Performed a commit and push in Visual Studio

Everything from Visual Studio works fine for me.

If you attempt to perform a fetch/push/pull from GitHub for Unity, you will have to setup a Personal Access Token with visualstudio.com. Otherwise, GitHub for Unity will freeze until you setup the authentication with your machine. All GitHub for Unity (and any other git client) is doing is calling the command line for you, so you will have to make sure fetch/push/pull work from the command line before attempting to use them from GitHub for Unity.

- [StackOverflow: Connect to git repository in TFS or Visual Studio Online](https://stackoverflow.com/questions/16835212/connect-to-git-repository-in-tfs-or-visual-studio-online?noredirect=1&lq=1)

I hope this all helps.
You can close the issue when you are ready.

**Comment by cyberangel67** [comment]
Yeah I even tried that and had issues doing that as well, the test project
in Unity was a test project so I could test it out. I'll play around with
it some more.

Btw what version of UnityGithub is the latest, I am sure I saw 1.0.3/1.0.4
in the source and yet when I open Unity it says 1.0.0. It's possible that I
may have selected skip this version when it promoted me to upgrade and I
don't see a way to upgrade it to test this out further.

And thanks I do appreciate the help. Even though I was really looking for
the same features that Visual Studio offers inside of Unity, but i guess
that is a long way down the road.




Regards,
Andrew Scott

**Comment by cyberangel67** [comment]
Well I don't know then.

I did exactly what you did and it wont sync for me.

I get this

Failed to push the branch to the remote repository. See the Output window for more details

and the output window says the same thing with the repository url, when I check the url the project was created but not files.

**Comment by StanleyGoldman** [comment]
The latest version of GitHub for Unity is 1.0.3,

I don't understand. When you get to sync in Visual Studio it should be coaxing you to push the project for the first time to a remote repo. How could it be pushing anywhere already?

**Comment by cyberangel67** [comment]
And I don't know why, but Unity GitHub won't update for me

**Comment by shana** [comment]
@cyberangel67 We can't fix Git issues regarding Visual Studio, for obvious reasons (this is GitHub for Unity, not Git for Visual Studio). If Visual Studio can't push to your git hosting provider (whether that is visualstudio.com or github.com), you'll have to sort that out in Visual Studio.

As for GitHub for Unity not updating, I've left instructions on how to manually update in your other issue #912. To be clear, GitHub for Unity doesn't auto update - it will give you a popup saying there is an update, and you will need to click on the button in the popup to download the update and then install it (by opening the downloaded package file while you have Unity running with your project open).
